### ISSUE TITLE
client-go: Add metrics into Informer

### ISSUE THREAD

**Issue by chenk008** [kubernetes/kubernetes#122067]
### What would you like to be added?

The informer should expose metrics about queue/reflector/eventHandler.

### Why is this needed?

1. The informer lacks of metric,  it is hard to known how many item in its queue/store. Add metrics for queue/store, it will help developers to find the number of pending deltas.

2. The informer create a RingGrowing pendingNotifications for every eventHandler. This RingGrowing will grow, but never shrink.  An informer has some eventHandlers, it is hard to distinguish which pendingNotifications cost a lot memory. Add metrics for eventHandler pendingNotification, it will help developers to distinguish which eventHandler is slow.

related issue:
https://github.com/kubernetes/kubernetes/issues/121662
https://github.com/kubernetes/kubernetes/issues/117123

**Comment by jiahuif** [comment]
Let's move the discussion to https://github.com/kubernetes/enhancements/issues/4346
/triage accepted

**Comment by nathanperkins** [comment]
We recently saw this log coming from controller-runtime, which I believe is backed by the standard informer code:

`"Unhandled Error" err="[sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108](http://sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108): Failed to watch *v1.Pod: failed to list *v1.Pod: Get \"https://<snip>/api/v1/namespaces/test-pods/pods?labelSelector=<snip>`

I'd ask for informer to expose these metrics so that we can properly alert when the informers backing our controllers are experiencing errors or not synced:

- Sync / Watch Errors
- Last successful sync time

**Comment by aojea** [comment]
The informers use client-go and other components like workqueue that already provide a lot of meaningful metrics 

https://github.com/kubernetes/client-go/blob/master/tools/metrics/metrics.go

https://github.com/kubernetes/kubernetes/blob/7dc4af6c8a67271e2146dc9c60d9f4b6adfbab3b/staging/src/k8s.io/client-go/util/workqueue/metrics.go

Specifically client-go will show error rate and responses already, have you checked that

**Comment by xigang** [comment]
> The informers use client-go and other components like workqueue that already provide a lot of meaningful metrics
> 
> https://github.com/kubernetes/client-go/blob/master/tools/metrics/metrics.go
> 
> https://github.com/kubernetes/kubernetes/blob/7dc4af6c8a67271e2146dc9c60d9f4b6adfbab3b/staging/src/k8s.io/client-go/util/workqueue/metrics.go
> 
> Specifically client-go will show error rate and responses already, have you checked that

@aojea We need a metric for the Informer DeltaFIFO data backlog, as this metric is very important.

**Comment by nathanperkins** [comment]
We are using the workqueue metrics, they are helpful for tracking the latency for work items waiting or processing in the queue. My understanding is that when the informer has errors which prevent it from watching objects (like the one posted above), then events are not going to be added to the workqueue and won't be reflected in the workqueue metrics.

I don't see the client-go metrics coming from controller-runtime by default. None of the metrics in that page jump out as useful in tracking whether the informer caches are recently updated or experiencing errors. I will look into that more.

**Comment by xigang** [comment]
Let's move the discussion: https://github.com/kubernetes/enhancements/issues/5187

**Comment by nathanperkins** [comment]
@xigang in that issue, the "Discussion Link" references this issue so that's why I discussed over here. Maybe we should update the description?

**Comment by xigang** [comment]
@nathanperkins I've implemented [ KEP 4346 ](https://github.com/kubernetes/kubernetes/pull/129160) Can you review the code?

**Comment by nathanperkins** [comment]
@xigang sorry, I'm not an owner or familiar with any of that code so I wouldn't be able to give a great review, other than standard golang readability advice.

**Comment by aojea** [comment]
We are conflating multiple things in this issue:
- informers are builing on top of  client-go and workqueue and tools/cache, both client-go and workqueue has metrics, @xigang has a KEP to expand the informer only metrics  https://github.com/kubernetes/kubernetes/pull/129160
- client-go already has metrics for detecting issues with network connections , other thing is people that use it through controller-runtime https://github.com/kubernetes/kubernetes/issues/127739

Metrics are registerd when you import the corresponding module in your component, you just add this line

https://github.com/kubernetes/kubernetes/blob/0b00454617397cea2e9e6ed75b232a4e1e9cfb6d/cmd/kube-controller-manager/controller-manager.go#L29

you can see some usage samples here to detect issues by checking latency per example https://github.com/kubernetes/kubernetes/pull/106911#issuecomment-1048021553

**Comment by nathanperkins** [comment]
>client-go already has metrics for detecting issues with network connections

My goal is to fire an alert when controller caches haven't synced in a significant amount of time so they are operating on stale data and not enqueuing events. It doesn't look like any of the client-go metrics measure that in a satisfying way.

Even the rest client metrics which could expose failed requests are limited to `host`, `verb`, and `code` labels which makes it difficult for me to target my alert on informer stability and ignore transient errors on singleton client requests.

For those reasons, I think an informer specific metric makes sense to expose the last sync time or the staleness duration.

>Metrics are registerd when you import the corresponding module in your component, you just add this line

Weirdly, I've added that import to my service and I didn't see any additional metrics. I'm not sure if GKE is filtering them out. Will have to look into that more.

**Comment by aojea** [comment]
> My goal is to fire an alert when controller caches haven't synced in a significant amount of time so they are operating on stale data and not enqueuing events

your controller must not start until the caches are synced, you can signal that externally with a ready flag ...

**Comment by nathanperkins** [comment]
>your controller must not start until the caches are synced, you can signal that externally with a ready flag ...

Yes, but this failure happened after the caches were already synced and the controller was already started. We saw ERROR logs but no metric. We had to provide our own [DefaultWatchErrorHandler](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/cache#Options) so that we can increment a metric for it.

Proper syncing is the most absolutely critical part of using an informer, I'm not sure why we would be hesitant to expose a metric for it. It is probably the most obviously useful metric that an informer could have as it represents the readiness of the informer itself.

Checking informer status during startup only is not enough for a component which is critical to the proper functioning of the controller.

**Comment by aojea** [comment]
> Yes, but this failure happened after the caches were already synced and the controller was already started. We saw ERROR logs but no metric. T

@nathanperkins you are explaining the symptom and coming to a conclusion but you are not explaining the root cause, what do you mean by proper syncing? the informer syncs once and the watches ... if the connection breaks it reconnect ... I'm not able to understand the problem hence I do not understand what metric do you need 

>  caches were already synced and the controller was already started.
> We saw ERROR logs but no metric
> Proper syncing is the most absolutely critical part of using an informer

**Comment by nathanperkins** [comment]
>you are explaining the symptom and coming to a conclusion but you are not explaining the root cause

In our case, the root cause was a transient issue in either gke-metadata-server or the control plane.

The root cause doesn't matter though, the part that matters is that the informer can be in a state where the watch is failed and not syncing properly. It is better to alert on the symptom (informer not synced) than the root cause. There could be many root causes and it's not reasonable to expect the user to know what the root causes could be and alert on those.

For example, we don't alert on root causes like image pull backoff or invalid flags, we alert on general pod readiness and deployment availability. We don't alert on mutex contention, we alert on API latency.

The informer should be the same way. We shouldn't have to craft alerts based on the underlying client experiencing errors, we should be able to alert on the general readiness of the informer, the fact that it is properly syncing and handling events.

>what do you mean by proper syncing?

I mean that the informer is synced properly using watches, so the caches are updating and the events are being handled.

>the informer syncs once and then watches ... if the connection breaks it reconnect

Yes, but there are conditions where it cannot reconnect for whatever reason. We want to be able to alert when that is happening.

**Comment by nathanperkins** [comment]
RCA is definitely useful, that's not what I'm saying. I mean that alerts should not focus on root causes, they should focus on exposing service degradation that needs to be addressed by an engineer. While/after addressing the alert, the engineer will do RCA and that's when having other, more detailed logs and metrics is helpful.

If we use alerts that focus on root causes, it's not ideal:

- We have to configure a lot more alerts to cover each of the causes.
- Causes are often lower level, and aren't always easy to understand and connect to the symptom programmatically.
- When you're alert on low level "causes", they're not always actually causing an issue. One example would be if you alerted on any client-go errors, but those errors were retried immediately and succeeded.
- If there is some service degradation with a root cause we haven't identified and alerted on, our service will be degraded and nobody will know to address it.

So, exposing some simple metric which summarizes the state of the informer as "ready" or "degraded" will help users make alerts to ensure the health of their system. I don't know enough about informer internals to know what would work, but my first thought is to pattern them off the metrics for container status:

- `informer_status_phase/gauge` (similar to `kube_pod_container_status_phase`) - always emits 1 with a label to show the status (like init, ready, error)
- `informer_status_ready/gauge` (similar to `kube_pod_container_status_ready`) - 1 when ready, 0 when initializing or when watch is experiencing errors and trying to reconnect
